apps/examples/messaging_sample/messaging_main.c

```diff
@@ -0,0 +1,262 @@
+	if (ret != OK) {
+		printf("Fail to sync send message.\n");
+		free(reply_param.msg);
+		return ERROR;
+	}
+
+	printf("Success to receive reply from receiver. : %s\n", (char *)reply_param.msg);
+	free(reply_param.msg);
+
+	return 0;
+}
+
+int nonblock_recv(int argc, FAR char *argv[])
+{
+	int ret;
+	msg_recv_param_t param;
+	msg_recv_data_t data;
+
+	param.type = MSG_RECV_TYPE_NONBLOCK;
+	param.cb_func = recv_callback;
+	data.msg = (char *)malloc(BUFFER_SIZE);
+	if (data.msg == NULL) {
+		printf("Fail to recv nonblock message : out of memory.\n");
+		return ERROR;
+	}
+	data.msglen = BUFFER_SIZE;
+
+	ret = messaging_recv(TEST1_PORT, &data, &param);
+	if (ret != OK) {
+		printf("Fail to receive with non-block mode.\n");
+		free(data.msg);
+		return ERROR;
+	}
+
+	free(data.msg);
+	return OK;
+}
+
+int block_recv(int argc, FAR char *argv[])
+{
+	int ret;
+	msg_recv_param_t recv_param;
+	msg_recv_data_t recv_data;
+	msg_reply_param_t reply_param;
+
+	recv_param.type = MSG_RECV_TYPE_BLOCK;
+	recv_data.msg = (char *)malloc(BUFFER_SIZE);
+	if (recv_data.msg == NULL) {
+		printf("Fail to receive, because out of memory.\n");
+	}
+	recv_data.msglen = BUFFER_SIZE;
+
+	ret = messaging_recv(TEST2_PORT, &recv_data, &recv_param);
+	if (ret != OK) {
+		printf("Fail to receive with block mode.\n");
+		free(recv_data.msg);
+		return ERROR;
+	}
+
+	printf("Success to receive with block mode in receiver. : %s\n", (char *)recv_data.msg);
+	printf("- Reply to sender %s data.\n", TEST2_REPLY);
+
+	reply_param.msg = TEST2_REPLY;
+	reply_param.msglen = sizeof(TEST2_REPLY);
+	ret = messaging_send_reply(TEST2_PORT, &reply_param);
+	if (ret != OK) {
+		printf("Fail to reply.\n");
+		free(recv_data.msg);
+		return ERROR;
+	}
+
+	free(recv_data.msg);
+	return OK;
+}
+
+void async_nonblock_messaging_test(void)
+{
+	int ret;
+	int sender_pid;
+
+	printf("\n=== Start the Async Non-block messaging test. ===\n");
+
+	sender_pid = task_create("async_send", TASK_PRIO, STACKSIZE, async_send, NULL);
+	if (sender_pid < 0) {
+		printf("Fail to create async_send task.\n");
+		return;
+	}
+
+	ret = task_create("nonblock_recv", TASK_PRIO, STACKSIZE, nonblock_recv, NULL);
+	if (ret < 0) {
+		task_delete(sender_pid);
+		printf("Fail to create nonblock_recv task.\n");
+		return;
+	}
+
+	/* Wait for finishing async_send and nonblock_recv tasks. */
+	sleep(1);
+
+	ret = messaging_unregister(TEST1_PORT);
+	if (ret != OK) {
+		printf("Fail to unregister TEST1_PORT.\n");
+		return;
+	}
+}
+
+void sync_block_messaging_test(void)
+{
+	int ret;
+	int sender_pid;
+
+	printf("\n=== Start the Sync Block messaging test. ===\n");
+
+	sender_pid = task_create("sync_send", TASK_PRIO, STACKSIZE, sync_send, NULL);
+	if (sender_pid < 0) {
+		printf("Fail to create sync_send task.\n");
+		return;
+	}
+
+	ret = task_create("block_recv", TASK_PRIO, STACKSIZE, block_recv, NULL);
+	if (ret < 0) {
+		free(sync_send_alloc);
+		task_delete(sender_pid);
+		printf("Fail to create block_recv task.\n");
+		return;
+	}
```

Comment:
No need to wait for other task to complete before exiting the main thread?

Reply:
Yes. It is okay.